kafka-client/src/main/java/dev/responsive/utils/SubPartitioner.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 Responsive Computing, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.utils;
+
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.streams.state.internals.Murmur3;
+
+/**
+ * {@code SubPartitioner} allows sub-partitioning a partition
+ * space into additional partitions such that there is a 1:N
+ * mapping of original partitions to new partitions (namely,
+ * if two events are in the same new partition they must have
+ * been in the same original partition, but the opposite is
+ * not true).
+ *
+ * <p>The algorithm used simply maps each original partition
+ * to {@code n} new partitions, beginning with {@code original
+ * * n} and ending with {@code original * n + n} (exclusive). To
+ * compute which new partition a key falls into, it applies a hash
+ * function and mods by {@code n}, adding that value to the base
+ * mapped partition.</p>
+ */
+public class SubPartitioner {
+
+  public static final SubPartitioner NO_SUBPARTITIONS = new SubPartitioner(1, k -> 0);
+
+  // ensure that the default sub-partitioning hasher is unlikely
+  // to be the same as the hasher that was used for the original
+  // partition scheme - if the hashers are the same, and the
+  // number of sub partitions is equal to the number of original
+  // partitions, all keys would be mapped to the same sub-partition
+  // and all other sub-partitions would be empty
+  private static final int SALT = 31;
```

Comment:
It might be good to record this in Cassandra and make sure it never changes. We could do it in a follow-up though.